### DISCUSSION TITLE
How to write a custom template rule

### DISCUSSION THREAD

**Discussion by dhcode** [angular-eslint/angular-eslint#849] [Q&A] [answered]
Is there a tutorial somewhere?

I tried to figure out how to write a custom rule for my templates.
But it seems like I can't reach required utilities, that are contained in the `@angular-eslint/eslint-plugin-template` package.

I can't import `@angular-eslint/eslint-plugin-template/dist/utils/create-eslint-rule`  in ` @angular-eslint/eslint-plugin-template@13.0.1`

I started writing a rule to forbid the use of `<a href="javascript:">` links.
```typescript
import {
  createESLintRule,
  getTemplateParserServices
} from '@angular-eslint/eslint-plugin-template/dist/utils/create-eslint-rule';
import { TmplAstElement } from '@angular-eslint/bundled-angular-compiler';
import { TSESLint } from '@typescript-eslint/experimental-utils';

export const RULE_NAME = 'no-javascript-links';
export type MessageIds = 'illegalHref';
export type Options = [];

export default createESLintRule<Options, MessageIds>({
  name: RULE_NAME,
  defaultOptions: [],
  meta: {
    type: 'problem',
    docs: {
      description: 'Usage of href="javascript:" is forbidden',
      recommended: 'error'
    },
    schema: [],
    messages: {
      illegalHref: 'Usage of href="javascript:" is forbidden'
    }
  },
  create(context: Readonly<TSESLint.RuleContext<MessageIds, Options>>) {
    const parserServices = getTemplateParserServices(context);

    return {
      'Element$1[name=/^(a)$/]'(node: TmplAstElement) {
        const href = node.attributes.find(({ name }) => name === 'href');
        if (href && href.value && href.value.match(/^javascript:/)) {
          const loc = parserServices.convertElementSourceSpanToLoc(context, node);

          context.report({
            loc,
            messageId: 'illegalHref',
            data: {
              element: node.name
            }
          });
        }
      }
    };
  }
});

```

Does anyone know how to solve this?

**Comment by reduckted** [comment]
I want to write my own custom rules too, so making `getTemplateParserServices` and `ParserServices` publicly accessible would be great, but I think you can work around it.

`createESLintRule` seems to be a wrapper to make it easier to produce documentation, so as long as you don't need documentation for your own rules, you should be able to ignore that function and just create the rule yourself.

And `getTemplateParserServices` doesn't appear to actually do anything important, other than verify that the parser exists on the context and then return it. You will need to redefine the [`ParserServices`](https://github.com/angular-eslint/angular-eslint/blob/2000c0c63f7b2c4d67c387cce4be8d039885da40/packages/eslint-plugin-template/src/utils/create-eslint-rule.ts#L41-L49) interface yourself, but `getTemplateParserServices` can be replaced with this `context.parserServices`.

I haven't tried your rule, but I have created a rule myself and it worked. I think this is all that you would need to change from your code above:

```diff
-import {
-  createESLintRule,
-  getTemplateParserServices
-} from '@angular-eslint/eslint-plugin-template/dist/utils/create-eslint-rule';
 import { TmplAstElement } from '@angular-eslint/bundled-angular-compiler';
 import { TSESLint } from '@typescript-eslint/experimental-utils';
 
 export const RULE_NAME = 'no-javascript-links';
 export type MessageIds = 'illegalHref';
 export type Options = [];
 
-export default createESLintRule<Options, MessageIds>({
+const RULE = {
   name: RULE_NAME,
   defaultOptions: [],
   meta: {
     type: 'problem',
     docs: {
       description: 'Usage of href="javascript:" is forbidden',
       recommended: 'error'
     },
     schema: [],
     messages: {
       illegalHref: 'Usage of href="javascript:" is forbidden'
     }
   },
   create(context: Readonly<TSESLint.RuleContext<MessageIds, Options>>) {
-    const parserServices = getTemplateParserServices(context);
+    const parserServices = (context as any).parserServices as ParserServices;
 
     return {
       'Element$1[name=/^(a)$/]'(node: TmplAstElement) {
         const href = node.attributes.find(({ name }) => name === 'href');
         if (href && href.value && href.value.match(/^javascript:/)) {
           const loc = parserServices.convertElementSourceSpanToLoc(context, node);
 
           context.report({
             loc,
             messageId: 'illegalHref',
             data: {
               element: node.name
             }
           });
         }
       }
     };
   }
-});
+};
+
+export default RULE;

+interface ParserServices {
+  convertNodeSourceSpanToLoc: (sourceSpan: ParseSourceSpan, ) => TSESTree.SourceLocation;
+  convertElementSourceSpanToLoc: (context: Readonly<TSESLint.RuleContext<string, readonly unknown[]>>, node: TmplAstElement, ) => TSESTree.SourceLocation;
+}
```

**Reply by jerone** [reply]
Pingback: https://github.com/angular-eslint/angular-eslint/issues/888

**Comment by jerone** [comment]
I started another discussion to try to get an overview of other Angular ESLint custom rules: https://github.com/angular-eslint/angular-eslint/discussions/1106

Besides being useful for projects developers, it might also be helpful for rule developers.

**Reply by pumano** [reply]
I totally agree with @jerone, because I currently try to make custom rule and got errors, while trying to do as original rules written. Just need to expose that methods properly

**Comment by JamesHenry** [comment, marked as answer]
I see it wasn't linked on this discussion yet, but I added a dedicated guide for this a while ago here: https://github.com/angular-eslint/angular-eslint/blob/77943ba54314f3414b326a96d9a519c2f314e3dd/docs/WRITING_CUSTOM_RULES.md
